Improve TerminationMessagePath to be more flexible

smarterclayton opened the pull request:
* Support `terminationMessagePolicy: fallbackToLogsOnError` which allows pod authors to get useful information from containers as per kubernetes/community#154
* Set an upper bound on the size of the termination message path or log output to prevent callers from DoSing the master
* Add tests for running as root, non-root, and for the new terminationMessagePolicy cases.

I set the limit to 4096 bytes, but this may be too high for large pod containers.  Probably need to set an absolute bound, i.e. max message size allowed is 20k total, and we truncate if we're above that limit.

Fixes #31839, #23569

```release-note
A new field `terminationMessagePolicy` has been added to containers that allows a user to request `FallbackToLogsOnError`, which will read from the container's logs to populate the termination message if the user does not write to the termination message log file.  The termination message file is now properly readable for end users and has a maximum size (4k bytes) to prevent abuse.  Each pod may have up to 12k bytes of termination messages before the contents of each will be truncated.
```

smarterclayton committed:
Limit the size of the termination log and allow log input

Enforce the following limits:

12kb for total message length in container status
4kb for the termination message path file
2kb or 80 lines (whichever is shorter) from the log on error

Fallback to log output if the user requests it.

smarterclayton committed:
bump:(github.com/armon/circbuf):bbbad097214e2918d8543d5201d12bfd7bca254d

smarterclayton committed:
Add TerminationMessagePolicy

smarterclayton committed:
generated: api

smarterclayton committed:
Add conformance tests for terminationMessage(Path|Policy)

Test root, non-root, success and message, failure and message.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39341)
<!-- Reviewable:end -->

smarterclayton committed:
Use chmod to bypass umask on termination log file

os.Create() will obey the umask which results in the file being 0644
when injected in the container.

smarterclayton commented:
@feiskyer you'll need to review the CRI changes (should be 1:1 with Docker, but maybe I missed something obvious in the structure of ReadLogs).  This should be no worse than Docker in performance for logs, but if tail reading is particularly inefficient we should discuss how to optimize it.

feiskyer commented:
@smarterclayton  Reviewing.

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -323,29 +334,68 @@ func makeUID() string {
-					message = string(data)
-				}
-				break
+			if mount.ContainerPath == nil || mount.HostPath == nil || *mount.ContainerPath != terminationMessagePath {
```
nit: replace with `if mount.GetContainerPath() != terminationMessagePath`

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -323,29 +334,68 @@ func makeUID() string {
+			if mount.ContainerPath == nil || mount.HostPath == nil || *mount.ContainerPath != terminationMessagePath {
+				continue
+			}
+			path := *mount.HostPath
```
nit: `mount.GetHostPath()`

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -323,29 +334,68 @@ func makeUID() string {
+
+// readLastBytesFromFile reads at most max bytes from the provided file or returns an error.
+// It returns true if it the file was longer than max.
+func readLastBytesFromFile(path string, max int64) ([]byte, bool, error) {
```
this is same with dockertools.readLastBytesFromFile, maybe move to `pkg/kubelet/util`?

smarterclayton replied:
Yeah

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -323,29 +334,68 @@ func makeUID() string {
-		if terminationMessagePath == "" {
-			return ""
-		}
+const maxMessageLength = 1024 * 4
```
nit: move const to the beginning of file.

Random-Liu replied:
Add comment for the const. Is it line number or byte number or something else.

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -323,29 +334,68 @@ func makeUID() string {
+// getTerminationMessage looks on the filesystem for the provided termination message path, returning a limited
+// amount of those bytes, or returns true if the logs should be checked.
+func getTerminationMessage(status *runtimeapi.ContainerStatus, terminationMessagePath string, fallbackToLogs bool) (string, bool) {
+	if len(terminationMessagePath) != 0 {
```
missing conditions of `if !kubeStatus.FinishedAt.IsZero() || kubeStatus.ExitCode != 0`

smarterclayton replied:
That's above now, in the status function, to be consistent with docker_manager.

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -391,13 +441,20 @@ func (m *kubeGenericRuntimeManager) getPodContainerStatuses(uid kubetypes.UID, n
-		// Use the termination message written by the application is not empty
-		if len(tMessage) != 0 {
-			cStatus.Message = tMessage
+			// TODO: how is OOMKill represented in CRI?
```
CRI doesn't define OOMKilled state clearly, but dockershim expose it by setting `ContainerStatus.Reason` to `OOMKilled`.

smarterclayton replied:
Ok, will translate.

feiskyer commented:
cc/ @Random-Liu maybe want another look

fejta commented:
@k8s-bot bazel test this
@k8s-bot unit test this
@k8s-bot verify test this

smarterclayton commented:
@k8s-bot test this issue #IGNORE

smarterclayton commented:
@kubernetes/sig-apps-misc for review of enhancements to termination message path to make it usable by a wider class of applications

smarterclayton commented:
Some feedback I got from our UI people - it's possible that this should be line based (up to the limit) for each container on fallback.  I.e. walk backwards a number of lines (10-100) trying to fill to the max size to read.  Number of lines of logs is a good correlation to history, and so might deliver more useful results to end users.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -323,29 +334,41 @@ func makeUID() string {
+	logOptions := &v1.PodLogOptions{}
+	// TODO: more efficient byte search option
+	value := int64(maxMessageLength)
+	logOptions.TailLines = &value
```
Do you want to set `LimitBytes` or `TailLines`? `LimitBytes` I guess.

smarterclayton replied:
LimitBytes is from the beginning of the logs, while TailLines is from the end.  If  we didn't want to avoid introducing a new API to LogOptions, I'd suggest TailBytes because it's fairly useful.  Unfortunately for here we have to get lines > bytes, then use a circular buffer to purge the extra bytes.

Random-Liu replied:
@smarterclayton I see. Thanks for the explanation.

Random-Liu commented on pkg/kubelet/dockertools/docker_manager.go:
```diff
@@ -509,6 +504,51 @@ func (dm *DockerManager) inspectContainer(id string, podName, podNamespace strin
 }
 
+func getTerminationMessage(c DockerInterface, iResult *dockertypes.ContainerJSON, terminationMessagePath string, fallbackToLogs bool) string {
+	const maxMessageLength = 1024 * 4
```
Add comment for this const.

smarterclayton commented on pkg/util/tail/tail.go:
```diff
@@ -0,0 +1,99 @@
+// * If n >= 0, return the beginning of last nth line.
+// Notice that if the last line is incomplete (no end-of-line), it will not be counted
+// as one line.
+func ReadAtMostLines(f io.ReadSeeker, n int64) (int64, error) {
```
Ah, totally missed that.  Yes, will rename.

lavalamp commented on pkg/api/v1/types.go:
```diff
@@ -1352,6 +1352,16 @@ const (
+	// FallbackToLogsOnErrorTerminationMessage means that if the termination message file is empty, the most
+	// recent contents of the container logs will be used as the termination message if the container exits
+	// with an error.
+	FallbackToLogsOnErrorTerminationMessage TerminationMessagePolicy = "FallbackToLogsOnError"
```
Hm, shouldn't this be `TerminationMessagePolicyReadLogsOnError` and I think you should add `TerminationMessagePolicyReadFilePath` or some such and make it the default. Then it would look intentional rather than something clearly tacked-on. :)

smarterclayton replied:
I admit I was trying to be clever - I didn't want to cause yet another new default field that would suddenly have to be set everywhere.  I agree that when we add policies we always add a default, but this might be the first time we've added a new string policy to change an existing field.  I also wanted to avoid setting this as long as possible so that it could be imposed by pod admission instead of on pod templates.

It would be more consistent to do so though, yes.

smarterclayton replied:
Going to rename to `File` and `FallbackToLogsOnError`.

lavalamp commented on pkg/api/v1/types.go:
```diff
@@ -1469,10 +1479,18 @@ type Container struct {
 	// Optional: Path at which the file to which the container's termination message
 	// will be written is mounted into the container's filesystem.
 	// Message written is intended to be brief final status, such as an assertion failure message.
+	// Will be truncated by the node if too large.
```
How should a user figure out how large is too large?

smarterclayton replied:
I left this vague in the doc until we had a discussion here about max size.

The current value is 4096 bytes per container.  That's potentially a lot.  I think 20-40 lines of backtrace in a normal log file would provide a fair amount of context, which is probably ~1600 bytes on average.  For termination files, we would want to accept more, since every byte belongs to something the user intends to send.  Overall, we can't accept more than probably 5k-20k per pod (the upper end of that range might be 5-10x the average pod size).

How about for simplicity sake we apply the following limits:

1. For reading logs we read 2k bytes or 40 lines, whichever is smaller (from the end of the logs)
2. For termination logs we allow up to 4k bytes (from the beginning of the file)
3. If all container messages take up more than 12k bytes, we truncate messages proportionally from the beginning of the message (rather than doing something more complicated which mixes beginning and end truncation).

smarterclayton replied:
Implemented this and added it to the godoc.

lavalamp commented:
I commented on the API changes. It's not clear that I'm the best approver for everything else though. The goal seems like a good one.

smarterclayton committed:
Fix tests after new defaulted value added

smarterclayton commented:
@bgrant0607 can you review the intent here and the API mechanism to ensure it aligns with your direction for TerminationMessagePath?

smarterclayton commented on pkg/util/tail/tail.go:
```diff
@@ -0,0 +1,99 @@
+	}
+	size := fi.Size()
+	if size == 0 {
+		return nil, true, nil
```
Yes, will change.

smarterclayton commented on test/e2e_node/runtime_conformance_test.go:
```diff
@@ -168,6 +168,128 @@ while true; do sleep 1; done
 				Expect(c.Delete()).To(Succeed())
 			})
+
+			It("should report termination message if TerminationMessagePath is set as non-root user [Conformance]", func() {
```
Switched to table driven and added a few more tests

bgrant0607 commented:
API LGTM

smarterclayton commented:
@feiskyer any other comments?

smarterclayton commented:
Any other comments from anyone?  Tests are passing, API approvals are there, comments addressed.

spxtr commented:
@k8s-bot bazel test this
https://github.com/kubernetes/test-infra/pull/1609

smarterclayton commented:
Got a verbal LGTM from @feiskyer, API from @lavalamp and @bgrant0607, if no other comments will rebase and apply label on monday.

bgrant0607 commented:
@smarterclayton 

I'm totally fine with this. I will predict, however, that people will ask to be able to configure the number of bytes/lines. In the future we should think about a hook that would enable someone to `grep | tail` the log after termination.

smarterclayton commented:
Hopefully godep isn't broken 

@k8s-bot verify test this

smarterclayton commented:
Green, applying label after trivial rebase.

bamb00 commented:
I think is a good idea would to write the error (for example: CrashLoopBackOff) to the container/pod log this will allow a log agent to collect and sent it to elasticsearch and get viewed in Kibana.  This would help the support team in determining the cause of failure, identify which host, application, and when.
